Add DllMap sample (#668)
dotnet/samples: core/extensions/DllMapDemo/ReadMe.md

# DllMap Demo

> This samples illustrates the use of NativeLibrary APIs to implement library-name mappings similar to [Mono](https://www.mono-project.com/) [Dllmap](http://www.mono-project.com/docs/advanced/pinvoke/dllmap/) feature.

Reviewer: ```suggestion
This sample illustrates the use of NativeLibrary APIs to implement library name mappings similar to the [Mono](https://www.mono-project.com/) [Dllmap](http://www.mono-project.com/docs/advanced/pinvoke/dllmap/) feature.
```

## NativeLibrary APIs

> .Net Core 3 provides a rich set of APIs to manage native libraries:

Reviewer: ```suggestion
.Net Core 3.0 provides a rich set of APIs to manage native libraries:
```

> - [NativeLibrary APIs](https://github.com/dotnet/corefx/blob/master/src/System.Runtime.InteropServices/ref/System.Runtime.InteropServices.cs#L728-L738): Perform operations on native libraries (such as `Load()`, `Free()`, get the address of an exported  symbol, etc.) in a platform-independent way from managed code.

Reviewer 2: The links should be to do documentation, not to the source code. (e.g. https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.nativelibrary?view=netcore-3.0)

Author: @Jkotas The Resolver callback is not in Preview 2, so there is no documentation for it yet. 
This sample is in anticipation for Preview 3. So, I can wait for the documentation to be updated to Preview 3 and then commit this sample.

Reviewer: ```suggestion
- [NativeLibrary APIs](xref:System.Runtime.InteropServices/NativeLibrary?displayProperty=nameWithType) perform operations on native libraries (such as `Load`, `Free`, get the address of an exported  symbol, etc.) in a platform-independent way from managed code.
```

> - DllImport Resolver callback:  Get a call-back for first-chance native library resolution using custom logic. 

Reviewer: ```suggestion
- DllImport Resolver callback:  Gets a callback for first-chance native library resolution using custom logic. 
```

> - [Native Library Resolve event](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.loader.assemblyloadcontext.resolvingunmanageddll?view=netcore-3.0): Get an event for last-chance native library resolution using custom logic.   

Reviewer: ```suggestion
- [Native Library Resolve event](xref:System.Runtime.Loader.AssemblyLoadContext.ResolvingUnmanagedDll?displayProperty=nameWithType>: Get an event for last-chance native library resolution using custom logic.   
```

## Library Mapping

> The above APIs can be used to implement custom native library resolution logic, including DllMap, as illustrated in this example. The sample demonstrates:

Reviewer: ```suggestion
These APIs can be used to implement custom native library resolution logic, including DllMap, as illustrated in this example. The sample demonstrates:
```

- An [app](Demo.cs) that pInvokes a method in `OldLib`, but runs in an environment where only [`NewLib`](NewLib.cpp) is available.
- The [XML file](Demo.xml) that maps the library name from `OldLib` to `NewLib`. 

> - The [Map](Map.cs) implementation, which parses the above mapping, and uses `NativeLibrary` APIs to load the correct library.

Reviewer: ```suggestion
- The [Map](Map.cs) implementation, which parses the above mapping and uses `NativeLibrary` APIs to load the correct library.
```
